File: flask_hashcash.js

function createCookie(name,value,days) {
    var expires = "";
    if (days) {
        var date = new Date();
        date.setTime(date.getTime() + (days*24*60*60*1000));
        expires = "; expires=" + date.toUTCString();
    }
    document.cookie = name + "=" + value + expires + "; path=/";
}

function readCookie(name) {
    var nameEQ = name + "=";
    var ca = document.cookie.split(';');
    for(var i=0;i < ca.length;i++) {
        var c = ca[i];
        while (c.charAt(0)==' ') c = c.substring(1,c.length);
        if (c.indexOf(nameEQ) == 0) return c.substring(nameEQ.length,c.length);
    }
    return null;
}

function hash(puzzle,solution){
    return Sha1.hash(puzzle.concat(solution)); // btw this is in binary
}

function findHashcashSolution(params = {}){
    default_params = {puzzle:null, difficulty:5, solution:0, callback:null};
    for(k in default_params)
        if(params[k] == undefined)
            params[k] = default_params[k];
    var puzzle = params.puzzle,
        difficulty = params.difficulty,
        solution = params.solution,
        callback = params.callback;   
    
    // read current puzzle from cookies
    if(puzzle == null)
        puzzle = readCookie('hashcash_puzzle');
    if(puzzle == null)
        return;
    // if you have cached solution no need to continue seach
    if(hash(puzzle, readCookie('hashcash_solution')) == Array(difficulty+1).join("0"))
        return;

    // check solution from params
    var h = hash(puzzle,solution);
    // if the hash starts with *difficulty*(default = 1) of zeros - solution verified
    if(h.substring(0,difficulty) == Array(difficulty+1).join("0")){
        createCookie('hashcash_solution', solution); //store it as cookie
        if(callback)
            callback.call(solution);
    }else
        // continuing search
        setTimeout(function(){
            params.solution += 1 // try next number
            findHashcashSolution(params);
        }, 0);
}

File: flask_hashcash.py
from functools import wraps
from flask import request, abort, session, jsonify, make_response
from string import hexdigits

import hashlib, time, random

def _generate_puzzle():
	return ''.join(random.choice(hexdigits) for _ in range(25)) 

def _wrong_solution_response():
	# you can do things here like ban user for some time
	# also consider saving time of puzzle generation and limit generations/sec per user
	return make_response('Wrong hashcash solution')

# main decorator
def validate_work(difficulty = 5,wrong_solution_response=_wrong_solution_response,\
				  generate_puzzle = _generate_puzzle):
	def decorator(f):
		@wraps(f)
		def decorated_function(*args, **kwargs):

			if 'hashcash_puzzle' not in session:
				# in case user doesn't have puzzle - generate it
				puzzle = generate_puzzle()
				session['hashcash_puzzle'] = puzzle # store in session
				resp = wrong_solution_response()
				resp.set_cookie('hashcash_puzzle', puzzle) # send out as cookie
				return resp

			puzzle = session['hashcash_puzzle']
			solution = request.cookies.get('hashcash_solution', '')

			# verify solution
			plain = '{}{}'.format(puzzle,solution)

			hash = hashlib.new('sha1',plain).digest()
			bin_hash = ''.join([bin(ord(c))[2:].zfill(8) for c in hash]) # convert to binary

			# check that the hash start whith *difficulty* of zeros 
			if bin_hash[:difficulty] != '0' * difficulty:
				resp = wrong_solution_response()
				resp.set_cookie('hashcash_puzzle', puzzle) # resend puzzle. just in case 
				return resp

			# at this point solution is verified 
			resp = f(*args, **kwargs) # endpoint response
			if type(resp) in [str,unicode]:
				resp = make_response(resp)

			# generate new puzzle
			puzzle = generate_puzzle()
			session['hashcash_puzzle'] = puzzle # store in session
			# stuff new puzzle into cookies
			resp.set_cookie('hashcash_puzzle', puzzle)
			resp.set_cookie('hashcash_solution', '', expires=0)
			return resp
		return decorated_function
	return decorator

